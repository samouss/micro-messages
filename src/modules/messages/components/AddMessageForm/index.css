.AddMessageForm {

}

.AddMessageForm__Row {
  display: flex;
  margin-bottom: 20px;
}

.AddMessageForm__Row:last-child {
  margin-bottom: 0;
  justify-content: flex-end;
}

.AddMessageForm__Row__Label {
  font-weight: 600;
  margin-right: 10px;
}

.AddMessageForm__Row__Textarea {
  font-size: 0.9rem;
  border: 1px solid #E5E5E5;
  width: 100%;
  height: 150px;
  padding: 10px;
  transition: border-color 0.25s;
  resize: none;
}

.AddMessageForm__Row__Textarea:focus {
  outline: none;
  border-color: #0169d9;
}
