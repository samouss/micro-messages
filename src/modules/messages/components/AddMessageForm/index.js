// @flow

import React, { Component } from 'react';
import Label from 'components/Label';
import { SubmitButton } from 'components/Button';
import { GroupInput, RadioInput } from 'components/Input';
import './index.css';

import type { MessageInput, Visibility } from '../../types';

export type Props = {
  onSubmit: MessageInput => Promise<void>,
};

export type State = MessageInput & {
  isLoading: boolean,
};

class AddMessageForm extends Component {
  props: Props;
  state: State;

  onChange: (Event & { currentTarget: HTMLInputElement }) => void;
  onSubmit: (Event) => void;

  constructor(props: Props) {
    super(props);

    this.state = {
      isLoading: false,
      body: '',
      visibility: 'public',
    };

    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  onChange(event: Event & { currentTarget: HTMLInputElement }) {
    this.setState({
      [event.currentTarget.name]: event.currentTarget.value,
    });
  }

  onSubmit(event: Event) {
    event.preventDefault();

    this.setState({ isLoading: true });

    return this.props.onSubmit({
      body: this.state.body,
      visibility: this.state.visibility,
    }).then(() => {
      this.setState({
        isLoading: false,
        body: '',
        visibility: 'public',
      });
    }).catch(() => {
      this.setState({ isLoading: false });
    });
  }

  render() {
    const { isLoading, body, visibility } = this.state;

    return (
      <form onSubmit={this.onSubmit}>
        <div styleName="AddMessageForm__Row">
          <span styleName="AddMessageForm__Row__Label">Visibilit√©:</span>
          <GroupInput>
            {(['public', 'private']: Array<Visibility>).map(current => (
              <RadioInput
                key={current}
                id={current}
                name="visibility"
                value={current}
                checked={current === visibility}
                onChange={this.onChange}
                required
              >
                <Label
                  type={current}
                />
              </RadioInput>
            ))}
          </GroupInput>
        </div>

        <div styleName="AddMessageForm__Row">
          <textarea
            name="body"
            styleName="AddMessageForm__Row__Textarea"
            placeholder="Entrer votre message..."
            onChange={this.onChange}
            value={body}
            required
          />
        </div>

        <div styleName="AddMessageForm__Row">
          <SubmitButton
            disabled={isLoading}
          >
            Envoyer
          </SubmitButton>
        </div>

      </form>
    );
  }
}

export default AddMessageForm;
