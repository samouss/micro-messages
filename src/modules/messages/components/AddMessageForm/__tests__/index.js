// @flow

import React from 'react';
import { shallow } from 'enzyme';
import AddMessageForm from '../index';

import type { Props, State } from '../index';

describe('<AddMessageForm />', () => {
  const defaultProps: Props = {
    onSubmit: () => Promise.resolve(),
  };

  it('expect to render the component correctly', () => {
    const props: Props = {
      ...defaultProps,
    };

    const component = shallow(
      <AddMessageForm
        {...props}
      />,
    );

    expect(component).toMatchSnapshot();
  });

  it('expect to render the component correctly with body', () => {
    const props: Props = {
      ...defaultProps,
    };

    const component = shallow(
      <AddMessageForm
        {...props}
      />,
    );

    // $FlowFixMe
    component.setState({ body: 'Some content' });

    expect(component).toMatchSnapshot();
  });

  describe('onChange', () => {
    it('expect to update the state on radio change', () => {
      const props: Props = {
        ...defaultProps,
      };

      const expectation = {
        before: {
          isLoading: false,
          visibility: 'public',
          body: '',
        },
        after: {
          isLoading: false,
          visibility: 'private',
          body: '',
        },
      };

      const component = shallow(
        <AddMessageForm
          {...props}
        />,
      );

      // $FlowFixMe
      expect(component.state()).toEqual(expectation.before);

      component
        .find('Radio')
        .first()
        .simulate('change', {
          currentTarget: {
            name: 'visibility',
            value: 'private',
          },
        });

      // $FlowFixMe
      expect(component.state()).toEqual(expectation.after);
    });

    it('expect to update the state on text change', () => {
      const props: Props = {
        ...defaultProps,
      };

      const expectation = {
        before: {
          isLoading: false,
          visibility: 'public',
          body: '',
        },
        after: {
          isLoading: false,
          visibility: 'public',
          body: 'Some text',
        },
      };

      const component = shallow(
        <AddMessageForm
          {...props}
        />,
      );

      // $FlowFixMe
      expect(component.state()).toEqual(expectation.before);

      component
        .find('[styleName="AddMessageForm__Row__Textarea"]')
        .simulate('change', {
          currentTarget: {
            name: 'body',
            value: 'Some text',
          },
        });

      // $FlowFixMe
      expect(component.state()).toEqual(expectation.after);
    });
  });

  describe('onSubmit', () => {
    it('expect to call onSubmit when form is submitted', () => {
      const props: Props = {
        ...defaultProps,
        onSubmit: jest.fn(() => Promise.resolve()),
      };

      const component = shallow(
        <AddMessageForm
          {...props}
        />,
      );

      const expectation = true;

      component
        .find('form')
        .simulate('submit', {
          preventDefault: () => {},
        });

      expect(props.onSubmit).toHaveBeenCalled();
      // $FlowFixMe
      expect(component.state().isLoading).toBe(expectation);
      expect(component).toMatchSnapshot();
    });

    it('expect to successfully call onSubmit', () => {
      expect.assertions(3);

      const preventDefault = jest.fn();

      const props: Props = {
        ...defaultProps,
        onSubmit: jest.fn(() => Promise.resolve()),
      };

      const nextState = {
        visibility: 'private',
        body: 'Some content',
      };

      const expectation: State = {
        isLoading: false,
        visibility: 'public',
        body: '',
      };

      const component = shallow(
        <AddMessageForm
          {...props}
        />,
      );

      // $FlowFixMe
      component.setState(nextState);

      // $FlowFixMe
      return component.instance().onSubmit({ preventDefault }).then(() => {
        expect(preventDefault).toHaveBeenCalled();
        expect(props.onSubmit).toHaveBeenCalledWith(nextState);
        // $FlowFixMe
        expect(component.state()).toEqual(expectation);
      });
    });

    it('expect to fail to call onSubmit', () => {
      expect.assertions(3);

      const preventDefault = jest.fn();

      const props: Props = {
        ...defaultProps,
        onSubmit: jest.fn(() => Promise.reject()),
      };

      const nextState = {
        visibility: 'private',
        body: 'Some content',
      };

      const expectation: State = {
        isLoading: false,
        visibility: 'private',
        body: 'Some content',
      };

      const component = shallow(
        <AddMessageForm
          {...props}
        />,
      );

      // $FlowFixMe
      component.setState(nextState);

      // $FlowFixMe
      return component.instance().onSubmit({ preventDefault }).then(() => {
        expect(preventDefault).toHaveBeenCalled();
        expect(props.onSubmit).toHaveBeenCalledWith(nextState);
        // $FlowFixMe
        expect(component.state()).toEqual(expectation);
      });
    });
  });
});
