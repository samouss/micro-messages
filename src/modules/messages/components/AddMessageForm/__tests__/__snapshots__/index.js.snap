// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddMessageForm /> expect to render the component correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    styleName="AddMessageForm__Row"
  >
    <span
      styleName="AddMessageForm__Row__Label"
    >
      Visibilité:
    </span>
    <Group>
      <Radio
        checked={true}
        id="public"
        name="visibility"
        onChange={[Function]}
        required={true}
        value="public"
      >
        <Label
          type="public"
        />
      </Radio>
      <Radio
        checked={false}
        id="private"
        name="visibility"
        onChange={[Function]}
        required={true}
        value="private"
      >
        <Label
          type="private"
        />
      </Radio>
    </Group>
  </div>
  <div
    styleName="AddMessageForm__Row"
  >
    <textarea
      name="body"
      onChange={[Function]}
      placeholder="Entrer votre message..."
      required={true}
      styleName="AddMessageForm__Row__Textarea"
      value=""
    />
  </div>
  <div
    styleName="AddMessageForm__Row"
  >
    <Submit
      disabled={false}
    >
      Envoyer
    </Submit>
  </div>
</form>
`;

exports[`<AddMessageForm /> expect to render the component correctly with body 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    styleName="AddMessageForm__Row"
  >
    <span
      styleName="AddMessageForm__Row__Label"
    >
      Visibilité:
    </span>
    <Group>
      <Radio
        checked={true}
        id="public"
        name="visibility"
        onChange={[Function]}
        required={true}
        value="public"
      >
        <Label
          type="public"
        />
      </Radio>
      <Radio
        checked={false}
        id="private"
        name="visibility"
        onChange={[Function]}
        required={true}
        value="private"
      >
        <Label
          type="private"
        />
      </Radio>
    </Group>
  </div>
  <div
    styleName="AddMessageForm__Row"
  >
    <textarea
      name="body"
      onChange={[Function]}
      placeholder="Entrer votre message..."
      required={true}
      styleName="AddMessageForm__Row__Textarea"
      value="Some content"
    />
  </div>
  <div
    styleName="AddMessageForm__Row"
  >
    <Submit
      disabled={false}
    >
      Envoyer
    </Submit>
  </div>
</form>
`;

exports[`<AddMessageForm /> onSubmit expect to call onSubmit when form is submitted 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    styleName="AddMessageForm__Row"
  >
    <span
      styleName="AddMessageForm__Row__Label"
    >
      Visibilité:
    </span>
    <Group>
      <Radio
        checked={true}
        id="public"
        name="visibility"
        onChange={[Function]}
        required={true}
        value="public"
      >
        <Label
          type="public"
        />
      </Radio>
      <Radio
        checked={false}
        id="private"
        name="visibility"
        onChange={[Function]}
        required={true}
        value="private"
      >
        <Label
          type="private"
        />
      </Radio>
    </Group>
  </div>
  <div
    styleName="AddMessageForm__Row"
  >
    <textarea
      name="body"
      onChange={[Function]}
      placeholder="Entrer votre message..."
      required={true}
      styleName="AddMessageForm__Row__Textarea"
      value=""
    />
  </div>
  <div
    styleName="AddMessageForm__Row"
  >
    <Submit
      disabled={true}
    >
      Envoyer
    </Submit>
  </div>
</form>
`;
